### ISSUE TITLE
sendState should behave asynchronously

### ISSUE THREAD

**Issue by cahirodoherty-learningpool** [adaptlearning/adapt-contrib-xapi#134]
### Subject of the issue/enhancement/features
We have a custom plugin that forces submit of multiple components in one functional batch
(In essence:
```
_.each(largeArrayOfComponentViews, (componentView) => {
        componentView.onSubmitClicked();
      });
```
)

When it comes to handling a large volume of components, we have found that this effectively creates a race condition and the state that comes back may actually not be the last state sent, thus cutting off some question values when the state is loaded again on subsequent views.

I believe setting an await on https://github.com/adaptlearning/adapt-contrib-xapi/blob/master/js/XAPI.js#L1082 should help us out in this regard

**Comment by cahirodoherty-learningpool** [comment]
@oliverfoster Any thoughts on this?

**Comment by oliverfoster** [comment]
The state is concurrently read in once and updated multiple times?

**Comment by cahirodoherty-learningpool** [comment]
Basically its a one-button submit for all components:

```_.each(this.componentViews, (componentView) => {
        componentView.onSubmitClicked();
    });```

**Comment by oliverfoster** [comment]
Yes, I don't understand where a state is being read in and in conflict. I see that submitting multiple questions at the same time would rapidly send statements.

**Comment by cahirodoherty-learningpool** [comment]
So based on examining the attached LRS, the component statements are registered with completion/correctness etc.
The problem lies with the overall state held in the LRS. So when the course is re-opened, the LRS state is loaded and many components appear to not have been completed. My hunch is that the large number of POST requests to update the course state are somehow being processed in the incorrect order and thus the last state to be saved, is not necessarily the last state intended to be processed.
From your comment about 'last updated', is this something that is in place by default on our submissions? I can't locate where that would be getting done

**Comment by oliverfoster** [comment]
I see, keeping a last updated date won't help you in this situation then, that would be added to the state when the client and sever need to be synched so that you can check which state is newer when handling copies.

A debounce for sending the lrs state would probably help here. Something like a half a second debounce, such that multiple updates can happen in quick succession in the client side and then a single update is sent to the server after the period has elapsed?

**Comment by cahirodoherty-learningpool** [comment]
Hi Oliver, yeah thats what we put in place to see us through and reports from the client seem to be positive. If thats the recommended solution and we aren't going to take any further action as a community then I'm happy to close this issue off now

**Comment by oliverfoster** [comment]
Could you push the debounce to the public repo if you haven't already?

**Comment by cahirodoherty-learningpool** [comment]
Its on the custom plugin side wrapped around the `onSubmitClicked` calls that I mentioned above. Were you talking about placing something in xAPI extension?
